src/docfx/lib/git/GitUtil.cs

```diff
@@ -0,0 +1,332 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Docs.Git;
+
+using FileMap = System.Collections.Generic.Dictionary<string, (string file, string parent)>;
+using TreeMap = System.Collections.Concurrent.ConcurrentDictionary<long, System.Collections.Generic.Dictionary<string, long>>;
+
+namespace Microsoft.Docs
+{
+    /// <summary>
+    /// Provide git operations
+    /// </summary>
+    public static class GitUtil
+    {
+        /// <summary>
+        /// Find git repo directory
+        /// </summary>
+        /// <param name="path">The git repo entry point</param>
+        /// <returns>The git repo root path</returns>
+        public static string FindRepo(string path)
+        {
+            var repo = path;
+            while (!string.IsNullOrEmpty(repo))
+            {
+                var gitPath = Path.Combine(repo, ".git");
+                if (Directory.Exists(gitPath) || File.Exists(gitPath) /* submodule */)
+                {
+                    return repo;
+                }
+                repo = Path.GetDirectoryName(repo);
+            }
+            return repo;
+        }
+
+        /// <summary>
+        /// Get git commits group by files
+        /// </summary>
+        /// <param name="repoPath">The git repo root path</param>
+        /// <param name="files">The collection of git repo files</param>
+        /// <param name="progress">The processing progress</param>
+        /// <returns>A collection of git commits</returns>
+        public static unsafe List<GitCommit>[] GetCommits(string repoPath, List<string> files, Action<int, int> progress = null)
+        {
+            if (string.IsNullOrEmpty(repoPath))
+            {
+                throw new ArgumentNullException(nameof(repoPath));
+            }
+
+            Debug.Assert(files.All(file => file == PathUtil.NormalizeFile(file)));
```

Comment:
>Debug.Assert(files.All(file => file == PathUtil.NormalizeFile(file))); [](start = 12, length = 70)

If `NormalizeFile` is used only in Debug.Assert, shall we exclude PathUtil in this PR and focus on GitUtility and here we check if it contains unwanted characters such as `..` or `\`?

Reply:
it not only used here, will add has invalid chars method later

---
In reply to: [180009088](https://github.com/dotnet/docfx/pull/2601#discussion_r180009088) [](ancestors = 180009088)